make/jdk/src/classes/build/tools/taglet/SealedGraph.java

```diff
@@ -0,0 +1,259 @@
+                + "<dd>"
+                + "<a class=\"sealed-graph\" href=\"" + imageFile + "\">"
+                + getImage(simpleTypeName, imageFile, thumbnailHeight, false)
+                + hoverImage
+                + "</a>"
+                + "</dd>";
+    }
+
+    private static final String VERTICAL_ALIGN = "vertical-align:top";
+    private static final String BORDER = "border: solid lightgray 1px;";
+
+    private String getImage(String typeName, String file, int height, boolean useBorder) {
+        return String.format("<img style=\"%s\" alt=\"Sealed class hierarchy graph for %s\" src=\"%s\"%s>",
+                useBorder ? BORDER + " " + VERTICAL_ALIGN : VERTICAL_ALIGN,
+                typeName,
+                file,
+                (height <= 0 ? "" : " height=\"" + height + "\""));
+    }
+
+    private static final class Renderer {
+
+        private Renderer() {
+        }
+
+        // Generates a graph in DOT format
+        static String graph(TypeElement rootClass, Set<String> exports) {
+            final State state = new State(rootClass);
+            traverse(state, rootClass, exports);
+            return state.render();
+        }
+
+        static void traverse(State state, TypeElement node, Set<String> exports) {
+            state.addNode(node);
+            for (TypeElement subNode : permittedSubclasses(node, exports)) {
+                if (isInPublicApi(node, exports) && isInPublicApi(subNode, exports)) {
+                    state.addEdge(node, subNode);
+                }
+                traverse(state, subNode, exports);
+            }
+        }
+
+        private static final class State {
+
+            private static final String LABEL = "label";
+            private static final String TOOLTIP = "tooltip";
+            private static final String STYLE = "style";
+
+            private final StringBuilder builder;
+
+            private final Map<String, Map<String, String>> nodeStyleMap;
+
+            public State(TypeElement rootNode) {
+                nodeStyleMap = new LinkedHashMap<>();
+                builder = new StringBuilder()
+                        .append("digraph G {")
+                        .append(lineSeparator())
+                        .append("  nodesep=0.500000;")
+                        .append(lineSeparator())
+                        .append("  ranksep=0.600000;")
+                        .append(lineSeparator())
+                        .append("  pencolor=transparent;")
+                        .append(lineSeparator())
+                        .append("  node [shape=plaintext, fontcolor=\"#e76f00\", fontname=\"DejaVuSans\", fontsize=12, margin=\".2,.2\"];")
+                        .append(lineSeparator())
+                        .append("  edge [penwidth=2, color=\"#999999\", arrowhead=open, arrowsize=1];")
+                        .append(lineSeparator())
+                        .append("  rankdir=\"BT\";")
+                        .append(lineSeparator());
```

Comment:
Note to future selves: if we come up with Yet Another Kind of Graph, we should consider creating a `DotBuilder` class to share/simplify the code.

Reply:
There is a dot-build for the ModuleGraph, but it makes a lot of assumptions about what kind of graph it is generating. I tried using that instead first, but it led to too much cludgy code, or too wide refactorings.